pkg/storegateway/bucket_stores.go

```diff
@@ -428,6 +457,49 @@ func (u *BucketStores) getOrCreateStore(userID string) (*store.BucketStore, erro
 	return bs, nil
 }
 
+// deleteLocalFilesForExcludedTenants removes local "sync" directories for tenants that are not included in the current
+// shard.
+func (u *BucketStores) deleteLocalFilesForExcludedTenants(includeUserIDs map[string]struct{}) {
+	files, err := ioutil.ReadDir(u.cfg.BucketStore.SyncDir)
+	if err != nil {
+		return
+	}
+
+	for _, f := range files {
+		if !f.IsDir() {
+			continue
+		}
+
+		userID := f.Name()
+		if _, included := includeUserIDs[userID]; included {
+			// Preserve directory for users owned by this shard.
+			continue
+		}
+
+		removed, bs := u.removeEmptyBucketStore(userID)
+		if bs != nil {
+			if !removed {
+				continue
+			}
+
+			err := bs.Close()
```

Comment:
[non-blocking] Why not moving this inside `removeEmptyBucketStore()`? It would allow us to not return the `*store.BucketStore`.

Reply:
Moved closing into `closeEmptyBucketStore`, which did complicate locking logic a bit, but nothing too bad. PTAL.